Add tests

wadjih-bencheikh18 opened the pull request:
closes : https://github.com/zakodium-oss/dom-to-html/issues/2
closes : https://github.com/zakodium-oss/dom-to-html/issues/3

lpatiny commented on src/index.ts:
```diff
@@ -3,8 +3,8 @@
  */
-export default async function domToHtml(dom: Element | null): Promise<string> {
-  if (dom === null || !dom.innerHTML) {
+export default function domToHtml(dom: Element): string {
```
Why is this function not async anymore ?

@wadjih-bencheikh18 I just commited a 'test' folder and it seems to me this code is not working.

@targos I would like that this test is available on the project home page and uses a build of the last version of the code but I don't know what is the best way to do this. Any suggestion ?

lpatiny replied:
Please also don't use 'default' so that we can easily add other methods

targos replied:
What is "the project home page" ?

lpatiny replied:
https://github.com/zakodium-oss/dom-to-html

lpatiny replied:
It could be a link from this page called 'Demo' for example.

targos replied:
I don't understand the point of doing manual testing when we are working on testing with Playwright.

wadjih-bencheikh18 replied:
> Why is this function not async anymore ?

I removed the async just to test the svg i'm still working on it

wadjih-bencheikh18 replied:
> I just commited a 'test' folder and it seems to me this code is not working.

i fixed the html test and full added test

lpatiny replied:
This library is expected to convert a page containing img, canvas and svg to encode them inline using 'base64' encoded images.

This means that the resulting html should only containing 'img' (so canvas and svg should be replaced) and the content should be base64 encoded.

I don't think that the test are currently checking this.

Also my mistakes that should be fixed:
1. I don't convert <img src=url> to inline img
2. I din't check the <canvas> conversion in the testcase

@wadjih-bencheikh18 could you check those points ?

wadjih-bencheikh18 replied:
> content should be base64 encoded.

the base64 encoding is not the same for the three devices chromium,firefox and webkit
so i used `expect(resultHtml).toBe(await result.evaluate((node) => node.innerHTML));` so resultHtml (code returned by domToHtml) compared to result content (html result rendred and compared to the test screenshot)

wadjih-bencheikh18 replied:
after many tests i found that the code is not working with svg tests so i reimplement domToHtml function

wadjih-bencheikh18 replied:
> I don't think that the test are currently checking this.

i'm checking that it's an `<img/>` and i think snapshots tests are enough to check that it's well encoded into base64

lpatiny reviewed (commented):
I would expose another function `copyDomToClipboard` that tries to copy a dom element to the clipboard and receives callback methods for 'error' and 'success'.

I think the test needs to be something like the 'test' folder I just commited. Load a page with jpg and svg and have a button to click on and check the html or the clipboard content is ok.

We need a page containing an img and a svg and a button 'click' button. When we click we should

lpatiny commented:
We added a failing testcase that 'matchSnapshot' and once the snapshot contains the base64 encode images (for the jpg and the svg converted to png) it is ok. (actually it pass but the snapshot is wrong).

For your other testcases you can also 'matchSnapshot'.

targos commented:
To make the snapshots work on GitHub CI, we need to do them with Linux. There is documentation here about how to do it using Docker: https://playwright.dev/docs/test-snapshots

wadjih-bencheikh18 commented:
@lpatiny i used toHaveScreenshot which ido the same work as toMatchSnapshot (if you perfer using toMatchSnapshot  i can change it)

`await expect(result).toHaveScreenshot( './src/__tests__/snapshots/canvas.png');`
<=>
`expect(await  result.screenshot()).toMatchSnapshot({name:'./src/__tests__/snapshots/canvas.png'});`

lpatiny commented on package.json:
```diff
@@ -18,7 +18,7 @@
     "eslint": "eslint src --cache",
     "eslint-fix": "npm run eslint -- --fix",
-    "playwright": "playwright test -c playwright.config.ts",
+    "playwright": "playwright test -c playwright.config.ts --update-snapshots",
```
We should not update snapshots during the tests ! They should be store in the project but we need to generate the snapshot for linux.

wadjih-bencheikh18 replied:
fixed thanks

stropitek commented on src/__tests__/domToHtml.test.tsx:
```diff
@@ -1,18 +1,76 @@
       </TestComponent>,
     );
+    const test = component.locator('#test');
+    await test.screenshot({ path: './src/__tests__/snapshots/basic.png' });
```
I'm not familiar with screenshot testing. What is the purpose of this line? Looking at playwright's documentation it seems you can just call `toHaveScreenshot` and it will create the image if it does not exist yet.

wadjih-bencheikh18 replied:
I used it to save screenshots but it's removed now thanks

wadjih-bencheikh18 replied:
Test page contain three sections
test => the original dom
html => text area to get html code generated with domToHtml
result=> the created dom with the html generated with domToHtml

so here i saved screenshot of test and compare it to result screenshot

lpatiny commented on src/__tests__/index.html:
```diff
@@ -0,0 +1,103 @@
+        const png = canvas.toDataURL('image/png');
+        return `<img src="${png}" />`;
+      }
+      function svgToHtml(svg) {
```
Actually you can not copy paste the resulting HTML in word if there is image/svg+xml. They have to be converted to png like we did for the other SVG (AFAIK).

So please also convert them.

lpatiny commented:
The goal is that you can copy paste the HTML to word and google docs and than all the images appear.

This means:
- inline SVG
- canvas
- <img src png
- <img src jpg
- <img src svg

Could you test the current status of the project with those 5 cases and copy / paste to google docs ?

lpatiny commented:
If you can share the google docs (read only mode) I would be convinced ;)

wadjih-bencheikh18 commented:
> If you can share the google docs (read only mode) I would be convinced ;)

i created this doc that contain test and the html result to test
https://docs.google.com/document/d/1TplZt64fC5kMalhnNTjTY8xsNjK3OuFkAEu3fhkXrpM/edit?usp=sharing

wadjih-bencheikh18 commented:
https://docs.google.com/document/d/1TplZt64fC5kMalhnNTjTY8xsNjK3OuFkAEu3fhkXrpM/edit?usp=sharing

@lpatiny the doc is updates with the clipboard please check it

stropitek reviewed (commented):
@wadjih-bencheikh18 there seems to still be quiet a few issues with the playwright screenshots, for example all the `svgImage` screenshots, or some of the `full` screenshots (and quiet many others)

Do you know why they are failing?

wadjih-bencheikh18 commented:
> Do you know why they are failing?

i fixed imgs dimensions bug
the only problem left is convert svg image in firefox that i'm tryin to solve it

i found a solution which is convert it into image/svg+xml base64 data, but it doesn't work when copy it in a doc using html clipboard

wadjih-bencheikh18 commented:
> > Do you know why they are failing?
> 
> i fixed imgs dimensions bug the only problem left is convert svg image in firefox that i'm tryin to solve it
> 
> i found a solution which is convert it into image/svg+xml base64 data, but it doesn't work when copy it in a doc using html clipboard so i converted it into image/png base64 data

lpatiny commented on src/index.ts:
```diff
@@ -3,64 +3,88 @@
+) {
+  return fetch(img.src)
+    .then((r) => r.blob())
+    .then(
```
Please use the await syntax
```
 const response = await fetch(img.src);
  const blob = await response.blob();

  return new Promise<void>((resolve) => {
    const reader = new FileReader();
    reader.onloadend = () => {
      const url = reader.result as string;
      const image = new Image();
      image.src = url;
      callback(image);
      resolve();
    };
    reader.readAsDataURL(blob);
  });
  ```

lpatiny commented on src/index.ts:
```diff
@@ -3,64 +3,88 @@
+    image.src = url;
+  });
+}
+async function imgToHtml(
```
Could you check that is works if the image is already base64 encoded ?

So if the html contains for example https://codepen.io/jamiekane/pen/YayWOa

I have the impression that fetch may fail and actually it should not try to replace this encoded image.

wadjih-bencheikh18 replied:
All results are saved in src/tests/domToHtml.test.tsx-snapshots

wadjih-bencheikh18 replied:
> I have the impression that fetch may fail and actually it should not try to replace this encoded image.

I'm actually replacing the domCopy elements with the encoded images
